### ISSUE TITLE
How can we reduce the size of the compiled file?

### ISSUE THREAD

**Issue by fuyw** [siboehm/lleaves#30]
Hello Simon,

I tried to compile a LightGBM model using both `LLeaves` and `TreeLite`. 

I found that the compiled `.so` file by `LLeaves` is ~80% larger than the files compiled by `TreeLite`.

I want ask that is it possible to reduce the size of the compiled `.so` file by `LLeaves`?

**Comment by siboehm** [comment]
Before I think about ways to make the binary smaller, may I ask why you want a smaller `.so` to begin with?

**Comment by fuyw** [comment]
Sure Simon, I totally understand that this may be a feature that we usually don't care.

In my application, I generate the models in one machine, and I need to upload them to a production machine. Sometimes the network is bad, so I would appreciate a model with a smaller size.

I am just writing to ask that are there any directions I might try to reduce the compiled model?

**Comment by siboehm** [comment]
Ok, that makes sense. In general you should expect that optimizing for codesize will slow down your predictions. Here's one thing that you could try: 
lleaves constructs an LLVM PassManager here, which runs the optimization passes: https://github.com/siboehm/lleaves/blob/359f1dc9e51171ce8495ba5a8eb74a524913bc66/lleaves/compiler/tree_compiler.py#L31-L33
You can add an extra line to set the `size_level`, see here for the docs: https://llvmlite.readthedocs.io/en/stable/user-guide/binding/optimization-passes.html#llvmlite.binding.PassManagerBuilder.size_level

That should bias the optimization passes towards generating code that is smaller, though I don't know how big the effect will be. You can test it, and if it works well I'll think of a way to adding a flag for size optimization to lleaves.

**Comment by siboehm** [comment]
I assume you're already exploiting the regular techniques for making uploading faster, like storing the model as a compressed binary and de-compressing it before loading it to memory.

**Comment by fuyw** [comment]
Thank you so much.

**Comment by siboehm** [comment]
How did you end up resolving your problem? Did any of my proposals work or did you switch to a different software?On 22. Oct 2022, at 13:38, fuyw ***@***.***> wrote:﻿
Closed #30 as completed.

**Comment by fuyw** [comment]
Hello Simon.

I have been a little busy lately, and I am so sorry for the late reply. 

I tried to add the `pmb.size_level = size_level` in the `tree_compiler.py`, where `size_level` is in [0, 1, 2].

```
 # Create optimizer 
 pmb = llvm.PassManagerBuilder() 
 pmb.opt_level = 3 
 pmb.size_level = size_level
````

I am not sure if I did it correctly. Because the compiled files have always the same sizes with different `size_level` values.

By the way, I noticed that the compiled file is in the type of `REL` (Relocatable file). So the `ldd` command won't display the dependencies on the complied model. 

I want to ask that is it possible that I output a complied file with `DYN` (dynamic shared library) type as in `.so`?
